File: src/netgraph/web/tour.py
"""The scratch inventories the guided tour edits, so the real one is untouched.

The tour in ``assets/tour.js`` does not mime its steps. It creates a device,
connects it to a neighbour, moves the document into another file, opens the
changes drawer on the resulting YAML, and undoes the lot — every one of those a
real batch through :mod:`netgraph.edit`, because a tour that only *described*
the write would prove nothing about the mapping between the picture and the
files, which is the single thing it exists to demonstrate.

So it is given files of its own. :meth:`Tours.open` copies the open inventory
into a temporary directory and puts a second, always-writable
:class:`~netgraph.web.session.EditingSession` over the copy; every route in
:mod:`netgraph.web.server` selects it with ``?scratch=<token>`` and answers from
it instead. Nothing about the real session changes — not its revision, not its
undo stack, not one byte on disk — and a read-only session can therefore take
the tour too, which is exactly the session somebody exploring is most likely to
have open.

Three properties this module is responsible for:

**Only inventory files are copied.** :func:`copy_inventory` walks the tree the
way :mod:`netgraph.loader.tree` reads it — YAML suffixes, no component starting
with ``.`` or ``_`` — plus ``netgraph.toml``, because the copy should render
with the same defaults the original does. A repository's history, its virtual
environments and its rendered SVGs are not part of an inventory and are not
copied.

**The copy is bounded.** :data:`MAX_FILES` and :data:`MAX_BYTES` stop a
mistyped ``netgraph web /`` from filling a disk, and the refusal names what it
counted so the number is actionable rather than mysterious.

**Nothing outlives the tab that asked for it.** A scratch is closed when the
tour finishes, when the page unloads (``navigator.sendBeacon``), when the server
stops, and, for the tab that crashed instead of doing any of those, when it has
gone :data:`TTL_SECONDS` without a request. :data:`MAX_SCRATCHES` caps how many
can exist at once.
"""

from __future__ import annotations

import shutil
import threading
import time
import uuid
from dataclasses import dataclass, field
from pathlib import Path
from tempfile import mkdtemp
from typing import Any, Final

from netgraph.loader.tree import YAML_SUFFIXES
from netgraph.render import IconTheme
from netgraph.web.session import EditingSession, SessionError

__all__ = [
    "CONFIG_NAME",
    "MAX_BYTES",
    "MAX_FILES",
    "MAX_SCRATCHES",
    "TOUR_END_PATH",
    "TOUR_PATH",
    "TTL_SECONDS",
    "Scratch",
    "Tours",
    "copy_inventory",
]

#: Start a tour: ``POST``, answered with the token every later request quotes.
TOUR_PATH: Final = "/api/tour"
#: End one and delete its files. ``POST``, and deliberately not a ``DELETE``:
#: ``navigator.sendBeacon`` — the only request a closing tab is allowed to make
#: — can only send a POST.
TOUR_END_PATH: Final = "/api/tour/end"

#: Copied beside the YAML so the scratch renders with the inventory's own
#: defaults; a tour whose diagrams look nothing like yours teaches the wrong
#: page. Read from the root only, which is the only place it is read from.
CONFIG_NAME: Final = "netgraph.toml"

#: Most files a scratch copy may hold. Well above the thousand-device tree
#: ``tools/bench_editor.py`` measures, and far below anything that would take
#: long enough to notice.
MAX_FILES: Final = 20_000
#: Most bytes of YAML a scratch copy may hold.
MAX_BYTES: Final = 64 * 1024 * 1024
#: Most tours that may be running at once. One per tab, and nobody has ten.
MAX_SCRATCHES: Final = 8
#: How long an untouched scratch survives. The backstop for a tab that closed
#: without saying so; every other path deletes it immediately.
TTL_SECONDS: Final = 3600.0


class TooLarge(SessionError):
    """The inventory is bigger than a scratch copy of it should be."""


@dataclass
class Scratch:
    """One tour's copy of an inventory, and the session over it."""

    #: What a request quotes in ``?scratch=`` to be answered from this copy.
    #: Unguessable, and not a credential: the server answers to this machine
    #: only, and a token names a temporary directory rather than a permission.
    token: str
    #: The temporary directory. Deleted by :meth:`close`.
    root: Path
    #: Always writable, whatever the session it was copied from allows.
    session: EditingSession
    #: The inventory it is a copy of, for the tour's first card to name.
    origin: Path
    #: An element the tour can cable its new device to, or ``""`` when the
    #: inventory has no device and the tour must create the far end itself.
    peer: str
    #: How many files were copied. Shown, because "a copy of your inventory" is
    #: a claim and a number is evidence.
    files: int
    #: Monotonic clock of the last request that named this scratch.
    touched: float = field(default_factory=time.monotonic)

    def to_dict(self) -> dict[str, Any]:
        """What ``POST /api/tour`` answers with."""
        return {
            "scratch": self.token,
            "root": str(self.root),
            "origin": str(self.origin),
            "peer": self.peer,
            "files": self.files,
        }

    def close(self) -> None:
        """Drop the session's streams and delete the copy."""
        self.session.close()
        # ``ignore_errors`` because this runs from a request thread, from a
        # beacon nobody is waiting for, and from server shutdown: a scratch that
        # cannot be removed — a file held open by a virus scanner on Windows is
        # the realistic case — is a temporary directory the operating system
        # will collect, not a reason to fail any of the three.
        shutil.rmtree(self.root, ignore_errors=True)


class Tours:
    """Every scratch copy this server is holding, by token.

    Shared by every request thread, so all of it is under one lock. Nothing here
    blocks for long: opening copies files, and everything else is a dict.
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._scratches: dict[str, Scratch] = {}

    def open(self, session: EditingSession, *, icons: IconTheme | None = None) -> Scratch:
        """Copy ``session``'s tree and put a writable session over the copy.

        Args:
            session: The inventory being toured. Read only — this never writes
                through it, bumps its revision or touches its history.
            icons: The theme the real session draws with, so the scratch draws
                the same way.

        Returns:
            The scratch, already registered under its token.

        Raises:
            TooLarge: The tree is beyond :data:`MAX_FILES` or :data:`MAX_BYTES`.
            SessionError: :data:`MAX_SCRATCHES` tours are already running.
        """
        self._expire()
        with self._lock:
            if len(self._scratches) >= MAX_SCRATCHES:
                raise SessionError(
                    f"{MAX_SCRATCHES} guided tours are already running on this server; "
                    "finish or close one before starting another"
                )
        root = Path(mkdtemp(prefix="netgraph-tour-"))
        try:
            files = copy_inventory(session.root, root)
            scratch = Scratch(
                token=uuid.uuid4().hex,
                root=root,
                session=EditingSession(root=root, writable=True, icons=icons),
                origin=Path(session.root),
                peer=_peer(session),
                files=files,
            )
        except BaseException:
            shutil.rmtree(root, ignore_errors=True)
            raise
        with self._lock:
            self._scratches[scratch.token] = scratch
        return scratch

    def get(self, token: str | None) -> Scratch | None:
        """The scratch a request named, or ``None`` when it named none.

        A token that is not one is ``None`` rather than a refusal: the request
        is then answered from the real session, which is read-only unless the
        command line said otherwise, so a stale token from a reloaded tab
        degrades into an ordinary request instead of an error nobody can act on.
        """
        if not token:
            return None
        with self._lock:
            scratch = self._scratches.get(token)
            if scratch is not None:
                scratch.touched = time.monotonic()
        return scratch

    def close(self, token: str | None) -> bool:
        """End one tour. ``True`` when there was one to end."""
        with self._lock:
            scratch = self._scratches.pop(token or "", None)
        if scratch is None:
            return False
        scratch.close()
        return True

    def close_all(self) -> None:
        """End every tour. Called when the server stops."""
        with self._lock:
            scratches = list(self._scratches.values())
            self._scratches.clear()
        for scratch in scratches:
            scratch.close()

    def _expire(self) -> None:
        """Drop scratches nothing has asked about for :data:`TTL_SECONDS`."""
        cutoff = time.monotonic() - TTL_SECONDS
        with self._lock:
            stale = [
                token for token, scratch in self._scratches.items() if scratch.touched < cutoff
            ]
            dropped = [self._scratches.pop(token) for token in stale]
        for scratch in dropped:
            scratch.close()

    def __len__(self) -> int:
        with self._lock:
            return len(self._scratches)


def copy_inventory(source: Path, destination: Path) -> int:
    """Copy the inventory documents under ``source`` into ``destination``.

    What the loader would read, and nothing else: files with a YAML suffix, no
    path component starting with ``.`` or ``_``, plus ``netgraph.toml`` at the
    root. Symbolic links are followed as files and never as directories, so a
    link out of the tree copies the one file it names rather than the world.

    Returns:
        How many files were copied.

    Raises:
        TooLarge: More than :data:`MAX_FILES` files or :data:`MAX_BYTES` bytes.
    """
    source = Path(source)
    destination = Path(destination)
    copied = 0
    total = 0
    for path in sorted(_documents(source)):
        relative = path.relative_to(source)
        target = destination / relative
        target.parent.mkdir(parents=True, exist_ok=True)
        size = path.stat().st_size
        copied += 1
        total += size
        if copied > MAX_FILES:
            raise TooLarge(
                f"the guided tour copies the inventory, and this one has more than "
                f"{MAX_FILES} documents; it is too big to tour"
            )
        if total > MAX_BYTES:
            raise TooLarge(
                f"the guided tour copies the inventory, and this one is over "
                f"{MAX_BYTES // (1024 * 1024)} MiB of YAML; it is too big to tour"
            )
        shutil.copyfile(path, target)
    config = source / CONFIG_NAME
    if config.is_file():
        shutil.copyfile(config, destination / CONFIG_NAME)
    return copied


def _documents(root: Path) -> list[Path]:
    """Every YAML file the loader would read below ``root``."""
    found: list[Path] = []
    stack = [root]
    while stack:
        directory = stack.pop()
        try:
            entries = list(directory.iterdir())
        except OSError:  # pragma: no cover - unreadable directory
            continue
        for entry in entries:
            if entry.name.startswith((".", "_")):
                continue
            if entry.is_dir() and not entry.is_symlink():
                stack.append(entry)
            elif entry.is_file() and entry.name.lower().endswith(YAML_SUFFIXES):
                found.append(entry)
    return found


def _peer(session: EditingSession) -> str:
    """A device the tour's new switch can be cabled to, or ``""``.

    The first device in load order, which is stable across runs of the same
    inventory and is therefore the same device in the screenshot and in the
    test. Chosen here rather than in the browser because the page sees drawn
    shapes and a filtered layer, and this needs a device that certainly has a
    ``spec.interfaces`` list to add a port to.

    An inventory with no device at all — a folder holding only cables, or an
    empty one — answers ``""``, and the tour creates both ends itself.
    """
    try:
        inventory = session.inventory()
    except Exception:  # pragma: no cover - a tree that will not load has no peer
        return ""
    for address in inventory.devices:
        return address
    return ""

File: src/netviz/web/assets/tour.js
/* The guided tour: sixty seconds that prove the diagram and the files are one
 * thing.
 *
 * A first-time visitor opens this page to a canvas, a file list and a palette
 * with four dozen entries, and nothing on screen says which of them is the
 * point. The point is that every shape is a document: create a device here and
 * a YAML file appears, move it and the file moves, undo and the bytes come
 * back. That is hard to say in a paragraph and obvious in a minute, so this
 * does it in a minute.
 *
 * It is not a mime. Every step posts a real batch to /api/ops, which goes
 * through netviz.edit exactly as `netviz edit create` does -- a tour that
 * faked its writes would demonstrate the one thing it is here to demonstrate
 * least well. What makes that safe is that the writes land somewhere else:
 * POST /api/tour copies the inventory into a temporary directory and opens a
 * second, writable session over the copy (netviz/web/tour.py), and while the
 * tour runs every request this page makes carries `?scratch=<token>` and is
 * answered from there.
 *
 * How the redirection is arranged, and why it is a page reload rather than a
 * re-attach: the token is put in sessionStorage and the page is reloaded, so
 * app.js and session.js boot against the scratch the ordinary way and there is
 * no second code path in which a page can be half attached to two sessions.
 * The wrapping of fetch and EventSource below happens at parse time, before
 * app.js has run, which is why this script is loaded before it.
 *
 * Because the copy is always writable, a read-only session can take the tour --
 * which is the session somebody who is just looking is most likely to have
 * open.
 *
 * Dependency-free, like the rest of this page.
 */

var netvizTour = (function () {
  "use strict";

  /** The running tour, in this tab only: a reload keeps it, a new tab does not. */
  var STORE = "netviz.tour";
  /** That the invitation has been answered, in this browser, for good. */
  var SEEN = "netviz.tour.seen";

  var START_PATH = "/api/tour";
  var END_PATH = "/api/tour/end";

  /** The element the tour creates, and the port it is cabled by. Fixed names,
   *  so the tour reads the same twice and a test can look for them. */
  var DEVICE = "sw-tour";
  var PORT = "eth0";
  var PEER_PORT = "eth-tour";
  var FILE = "tour/sw-tour.yaml";
  /** Where the tour's device is moved to, and the far end when the inventory
   *  has no device to cable to. */
  var LONE_PEER = "pc-tour";

  /** { token, origin, root, files, peer } while a tour is running. */
  var running = read();
  var host = null;
  var panel = null;
  var entry = null;
  var spot = null;
  var index = 0;
  /** How many batches have been applied, and therefore how many undos it takes
   *  to put the copy back exactly as it was found. */
  var batches = 0;
  var busy = false;

  /* --------------------------------------------------- the redirection */

  /** This tab's tour, or null. Unreadable storage is "no tour", not an error:
   *  a page that will not boot because of a private-browsing quirk is worse
   *  than one that simply does not offer the tour. */
  function read() {
    try {
      var raw = window.sessionStorage.getItem(STORE);
      return raw ? JSON.parse(raw) : null;
    } catch (error) {
      return null;
    }
  }

  function write(value) {
    try {
      if (value) { window.sessionStorage.setItem(STORE, JSON.stringify(value)); }
      else { window.sessionStorage.removeItem(STORE); }
    } catch (error) { /* nothing to do about it, and nothing to break */ }
  }

  function seen() {
    try { return window.localStorage.getItem(SEEN) === "yes"; }
    catch (error) { return true; }
  }

  function markSeen() {
    try { window.localStorage.setItem(SEEN, "yes"); }
    catch (error) { /* the invitation will come back; that is all */ }
  }

  /** Send this request to the scratch instead of the tree.
   *
   * Only our own routes, only while a tour is running, and never the two routes
   * that start and end one -- those are about the real session by definition.
   */
  function redirect(url) {
    var text = String(url);
    if (!running || text.indexOf("/api/") !== 0) { return text; }
    if (text.indexOf(START_PATH) === 0) { return text; }
    return text + (text.indexOf("?") === -1 ? "?" : "&")
      + "scratch=" + encodeURIComponent(running.token);
  }

  if (running) {
    var realFetch = window.fetch.bind(window);
    window.fetch = function (input, init) {
      return realFetch(typeof input === "string" ? redirect(input) : input, init);
    };
    if (window.EventSource) {
      var RealSource = window.EventSource;
      var Wrapped = function (url, config) { return new RealSource(redirect(url), config); };
      Wrapped.prototype = RealSource.prototype;
      window.EventSource = Wrapped;
    }
    /* The tab that is closed mid-tour: without this the copy sits in the
     * temporary directory until the server stops or the hour is up. */
    window.addEventListener("pagehide", function () {
      if (!running || !navigator.sendBeacon) { return; }
      navigator.sendBeacon(
        END_PATH + "?scratch=" + encodeURIComponent(running.token), new Blob([], {})
      );
    });
  }

  /* --------------------------------------------------------- the steps */

  /** One card. `run` is the edit it makes, and is what the Next button waits
   *  for; a card without one is prose. */
  function steps() {
    var peer = (running && running.peer) || "";
    var far = peer || LONE_PEER;
    return [
      {
        id: "welcome",
        title: "A sixty-second tour",
        body: function () {
          return "This page draws an inventory of YAML files, and every shape in it "
            + "is a document. The next five steps create a device, cable it up, move "
            + "its document to another file, show you the YAML that changed, and put "
            + "it all back.";
        },
        note: function () {
          return "Working on a copy of " + (running.files || 0) + " file"
            + (running.files === 1 ? "" : "s") + " made a moment ago. Your inventory is "
            + "not open for writing here at all.";
        },
        spot: "#files"
      },
      {
        id: "create",
        title: "Create a device",
        body: function () {
          return "One batch through the same write path <code>netviz edit create</code> "
            + "uses. Watch the file list on the left: a document has to be written "
            + "for the shape to exist, and netviz chooses where by looking at where "
            + "the inventory already puts things like it.";
        },
        spot: "#file-list",
        run: function () {
          return apply([{
            op: "create",
            kind: "switch",
            name: DEVICE,
            namespace: "",
            spec: { interfaces: [{ name: PORT, type: "ethernet" }] }
          }], "created switch " + DEVICE);
        },
        after: function () { return "created " + DEVICE + ", and the file that declares it"; }
      },
      {
        id: "connect",
        title: "Cable it to " + far,
        body: function () {
          return peer
            ? "A port on <code>" + esc(peer) + "</code>, and a cable between the two. "
              + "Two operations, one batch — so it is also one thing to undo. The link "
              + "appears in the diagram because a <code>kind: cable</code> document now "
              + "says it is there."
            : "This inventory has no device to cable to, so the tour creates the far "
              + "end as well. Three operations, one batch — so it is also one thing to "
              + "undo.";
        },
        spot: "#viewport",
        run: function () {
          var list = peer
            ? [{
                op: "add-interface",
                address: peer,
                interface: { name: PEER_PORT, type: "ethernet" }
              }]
            : [{
                op: "create",
                kind: "computer",
                name: LONE_PEER,
                namespace: "",
                spec: { interfaces: [{ name: PEER_PORT, type: "ethernet" }] }
              }];
          list.push({ op: "connect", a: DEVICE + ":" + PORT, b: far + ":" + PEER_PORT });
          return apply(list, "connected " + DEVICE + " to " + far);
        },
        after: function () { return "cabled " + DEVICE + " to " + far; }
      },
      {
        id: "move",
        title: "Move its document",
        body: function () {
          return "<code>" + esc(DEVICE) + "</code> does not change: the same element, "
            + "the same cable, the same diagram. Only the file it is declared in moves, "
            + "to <code>" + esc(FILE) + "</code>, which did not exist a second ago. "
            + "This is what it means for the layout of the tree to be yours and not "
            + "netviz's.";
        },
        spot: "#file-list",
        run: function () {
          return apply(
            [{ op: "move", address: DEVICE, file: FILE }],
            "moved " + DEVICE + " to " + FILE
          );
        },
        after: function () { return "moved the document to " + FILE; }
      },
      {
        id: "diff",
        title: "The YAML that changed",
        body: function () {
          return "Every gesture is logged with the unified diff it produced, and with "
            + "the <code>netviz edit</code> command that would repeat it from a "
            + "script. The diagram behind is now drawn as a diff too: what was added "
            + "is marked <span class=\"tour-sigil\">+</span>.";
        },
        spot: "#changes",
        run: function () {
          netvizSession.showChanges(true);
          return Promise.resolve(true);
        },
        after: function () {
          return batches + " change" + (batches === 1 ? "" : "s")
            + ", with the YAML each one wrote";
        }
      },
      {
        id: "undo",
        title: "Undo the lot",
        body: function () {
          return batches + " batch" + (batches === 1 ? "" : "es") + ", "
            + batches + " undo" + (batches === 1 ? "" : "s")
            + " — the history is the server's, so it survives a reload and is the "
            + "same history <code>Ctrl-Z</code> walks. "
            + "When it finishes, the copy is byte for byte what it was when the tour "
            + "started.";
        },
        spot: "#undo",
        run: function () {
          netvizSession.showChanges(false);
          return undoAll();
        },
        after: function () { return "undone; the files are back as they were"; }
      },
      {
        id: "done",
        title: "That is the whole idea",
        body: function () {
          return "The picture and the files are one thing, in both directions: edit "
            + "the YAML on the left and the diagram follows. Everything the tour did "
            + "is in the command palette — <kbd>Ctrl-K</kbd> — and everything the "
            + "palette does is a <code>netviz edit</code> command you can run from a "
            + "script. Press <kbd>?</kbd> for the keys.";
        },
        note: function () {
          return "Finishing deletes the copy and puts you back on "
            + (running.origin || "your inventory") + ".";
        },
        spot: "#commands",
        last: true
      }
    ];
  }

  /* -------------------------------------------------------- the driving */

  /** Post a batch to the scratch and count it, so undo knows how far to go. */
  function apply(list, said) {
    return netvizSession.ops(list, said).then(function (result) {
      batches += 1;
      return result;
    });
  }

  /** Undo every batch this tour applied, oldest last, one at a time. */
  function undoAll() {
    if (batches <= 0) { return Promise.resolve(true); }
    return netvizSession.step("undo").then(function (moved) {
      if (!moved) { batches = 0; return false; }
      batches -= 1;
      return undoAll();
    });
  }

  /* ---------------------------------------------------------- the panel */

  function esc(text) {
    var node = document.createElement("span");
    node.textContent = String(text === undefined || text === null ? "" : text);
    return node.innerHTML;
  }

  function build() {
    var node = document.createElement("aside");
    node.className = "tour";
    node.id = "tour";
    node.setAttribute("role", "dialog");
    node.setAttribute("aria-labelledby", "tour-title");
    node.setAttribute("aria-describedby", "tour-body");
    node.innerHTML =
      '<p class="tour-progress" id="tour-progress"></p>'
      + '<h2 id="tour-title"></h2>'
      + '<div id="tour-body" class="tour-body"></div>'
      + '<p id="tour-note" class="tour-note"></p>'
      + '<p id="tour-outcome" class="tour-note" role="status"></p>'
      + '<p class="tour-actions">'
      + '<button type="button" id="tour-next" class="tour-next"></button>'
      + '<button type="button" id="tour-skip" class="ghost">Skip <kbd>Esc</kbd></button>'
      + "</p>"
      + '<p class="tour-safe" id="tour-safe"></p>';
    node.querySelector("#tour-next").addEventListener("click", function () { advance(); });
    node.querySelector("#tour-skip").addEventListener("click", function () { finish(); });
    return node;
  }

  /** Outline the thing the current card is about, and nothing else. */
  function highlight(selector) {
    if (spot) { spot.classList.remove("tour-spot"); spot = null; }
    if (!selector) { return; }
    var target = document.querySelector(selector);
    if (!target) { return; }
    target.classList.add("tour-spot");
    spot = target;
  }

  function paint() {
    var list = steps();
    var step = list[index];
    panel.querySelector("#tour-progress").textContent =
      "Step " + (index + 1) + " of " + list.length;
    panel.querySelector("#tour-title").textContent = step.title;
    panel.querySelector("#tour-body").innerHTML = step.body();
    panel.querySelector("#tour-note").innerHTML = step.note ? step.note() : "";
    var outcome = panel.querySelector("#tour-outcome");
    outcome.className = "tour-note";
    outcome.textContent = "";
    var next = panel.querySelector("#tour-next");
    next.disabled = false;
    next.innerHTML = step.last
      ? 'Finish <kbd>&crarr;</kbd>'
      : (step.run ? "Do it" : "Next") + ' <kbd>&crarr;</kbd>';
    panel.querySelector("#tour-safe").innerHTML =
      "Editing a scratch copy in <code>" + esc(running.root) + "</code>. Nothing "
      + "under <code>" + esc(running.origin) + "</code> is written.";
    highlight(step.spot);
    netvizA11y.announce(step.title + ". " + panel.querySelector("#tour-body").textContent, false);
  }

  /** Do this card's edit, say what happened, and move on to the next. */
  function advance() {
    if (busy) { return; }
    var list = steps();
    var step = list[index];
    if (step.last) { finish(); return; }
    var next = panel.querySelector("#tour-next");
    if (!step.run) { index += 1; paint(); return; }
    busy = true;
    next.disabled = true;
    next.textContent = "working…";
    Promise.resolve()
      .then(step.run)
      .then(function () {
        busy = false;
        index += 1;
        paint();
        if (step.after) {
          var outcome = panel.querySelector("#tour-outcome");
          outcome.className = "tour-note done";
          outcome.textContent = step.after();
          netvizA11y.announce(step.after(), false);
        }
      })
      .catch(function (error) {
        // A refused edit is worth showing rather than hiding: this is a real
        // write path and it is allowed to say no. The tour goes on to the next
        // card, because a stuck tour is worse than an incomplete one.
        busy = false;
        index += 1;
        paint();
        var outcome = panel.querySelector("#tour-outcome");
        outcome.className = "tour-note failed";
        outcome.textContent =
          "that step was refused: " + String((error && error.message) || error);
      });
  }

  /** End the tour: drop the copy, forget the token, and reload onto the tree. */
  function finish() {
    var token = running && running.token;
    markSeen();
    write(null);
    highlight(null);
    if (entry) { netvizKeys.closeOverlay(entry); entry = null; }
    var done = function () { window.location.reload(); };
    if (!token) { done(); return; }
    // Deliberately not through the wrapped fetch's redirect: the token is the
    // body of this request, not the session it is answered from.
    window.fetch(END_PATH + "?scratch=" + encodeURIComponent(token), {
      method: "POST",
      cache: "no-store"
    }).then(done, done);
  }

  /* ------------------------------------------------------ the invitation */

  /** The one thing a first-time visitor is shown without asking. */
  function invite() {
    var node = document.createElement("aside");
    node.className = "tour tour-invite";
    node.id = "tour-invite";
    node.setAttribute("role", "dialog");
    node.setAttribute("aria-labelledby", "tour-invite-title");
    node.innerHTML =
      '<h2 id="tour-invite-title">First time here?</h2>'
      + '<div class="tour-body">Sixty seconds that create a device, cable it up, move '
      + "its document and undo the lot — on a throwaway copy of this inventory, so "
      + "your files are not touched.</div>"
      + '<p class="tour-actions">'
      + '<button type="button" id="tour-take" class="tour-next">Take the tour <kbd>&crarr;</kbd></button>'
      + '<button type="button" id="tour-not-now" class="ghost">No thanks <kbd>Esc</kbd></button>'
      + "</p>";
    var handle = null;
    var shut = function () {
      markSeen();
      netvizKeys.closeOverlay(handle);
    };
    var take = function () { shut(); start(); };
    node.querySelector("#tour-take").addEventListener("click", take);
    node.querySelector("#tour-not-now").addEventListener("click", shut);
    handle = netvizKeys.overlay(node, {
      focus: function () { node.querySelector("#tour-take").focus(); },
      close: shut,
      onKey: function (chord, event) {
        if (chord === "Escape") { event.preventDefault(); shut(); return true; }
        if (chord === "Enter") { event.preventDefault(); take(); return true; }
        return false;
      }
    });
  }

  /* ------------------------------------------------------------ starting */

  /** Ask for a scratch copy, then reload onto it. */
  function start() {
    if (running) { return Promise.resolve(true); }
    return window.fetch(START_PATH, { method: "POST", cache: "no-store" })
      .then(function (response) {
        return response.json().then(function (body) {
          if (!response.ok) { throw new Error(body.message || response.statusText); }
          return body;
        });
      })
      .then(function (body) {
        write({
          token: body.scratch,
          root: body.root,
          origin: body.origin,
          peer: body.peer,
          files: body.files
        });
        window.location.reload();
        return true;
      })
      .catch(function (error) {
        if (host) { host.toast("the tour could not start: " + error.message, "error"); }
        return false;
      });
  }

  /** Put the panel up for a tour this page has already reloaded into. */
  function resume() {
    panel = build();
    index = 0;
    batches = 0;
    entry = netvizKeys.overlay(panel, {
      focus: function () { panel.querySelector("#tour-next").focus(); },
      close: finish,
      onKey: function (chord, event) {
        if (chord === "Escape") { event.preventDefault(); finish(); return true; }
        if (chord === "Enter" || chord === "ArrowRight") {
          // Not while the focus is on Skip: Enter there means Skip.
          if (chord === "Enter" && document.activeElement
              && document.activeElement.id === "tour-skip") {
            return false;
          }
          event.preventDefault();
          advance();
          return true;
        }
        // Everything else still reaches the page underneath, which is the
        // difference between a tour and a hostage situation.
        return false;
      }
    });
    paint();
  }

  /* --------------------------------------------------------- the command */

  function defineCommands(bridge) {
    host = bridge;
    netvizKeys.define("tour", {
      run: function () {
        if (running) { netvizA11y.announce("the tour is already running", false); return; }
        start();
      }
    });
  }

  /** Called once, after the page knows which face it is and which session
   *  answered it.
   *
   * The check on `state.scratch` is the load-bearing line in this file. A token
   * the server no longer has -- it restarted, or the hour expired -- is
   * deliberately *not* a refusal: every route answers such a request from the
   * real tree instead, so that a reloaded tab gets a working page rather than a
   * dead one. Which means a tour that resumed without looking would drive its
   * create, connect and move against the inventory it exists to protect. So it
   * resumes only when the server confirms the copy is the thing answering.
   */
  function boot(mode, state) {
    if (running) {
      if (mode !== "session" || !state || state.scratch !== running.token) {
        write(null);
        window.location.reload();
        return false;
      }
      resume();
      return true;
    }
    if (mode === "session" && !seen()) { invite(); return true; }
    return false;
  }

  return {
    defineCommands: defineCommands,
    boot: boot,
    start: start,
    finish: finish,
    /** For tests: what this page thinks it is editing. */
    scratch: function () { return running ? running.token : ""; },
    isRunning: function () { return !!running; }
  };
})();
